zephyr/lib/regions_mm.c

```diff
@@ -341,101 +455,15 @@ void *vmh_alloc(struct vmh_heap *heap, uint32_t alloc_size)
-	 * done in the allocator section that is in CONFIG_MM_DRV_PAGE_SIZE chunk.
-	 * Start by getting start of allocation
-	 * ptr aligned to physical pages boundary
-	 */
-	phys_aligned_ptr = ALIGN_DOWN((uintptr_t)ptr, CONFIG_MM_DRV_PAGE_SIZE);
-	phys_aligned_alloc_end = ALIGN_UP((uintptr_t)ptr
-		+ alloc_size, CONFIG_MM_DRV_PAGE_SIZE);
-
-	/* To check if there was any allocation done before on block we need
-	 * to perform check operations however
-	 * we just made allocation there. We will clear it
-	 * from bit array of allocator for ease of operation.
-	 * RFC if someone knows a better way to handle it in bit array API.
-	 */
-
-	/* Clear bits of the allocation that was done */
-	sys_bitarray_clear_region(heap->physical_blocks_allocators[mem_block_iterator]->bitmap,
-		block_count, allocation_bitarray_position);
-
-	/* Now that we cleared the allocation we just made we can check
-	 * if there was any other allocation in the space that would need to be
-	 * mapped. Since this API is only one that uses mapping in specified address
-	 * ranges we can assume that we made mapping for the space if it has any previous
-	 * allocations. We check that iterating over all physical allocation blocks
-	 */
-	physical_block_count = (phys_aligned_alloc_end
-		- phys_aligned_ptr) / CONFIG_MM_DRV_PAGE_SIZE;
-
-	for (i = 0; i < physical_block_count; i++) {
-
-		phys_block_ptr = phys_aligned_ptr + i * CONFIG_MM_DRV_PAGE_SIZE;
-
-		check_offset = phys_block_ptr
-			- (uintptr_t)heap->physical_blocks_allocators[mem_block_iterator]->buffer;
-		check_position = check_offset / block_size;
-
-		check_size = CONFIG_MM_DRV_PAGE_SIZE / block_size;
-
-		/* We are now sure that if the allocation bit array between
-		 * block_realigned_ptr and block_realigned_alloc_end shows us if we already had
-		 * any allocations there and by our logic if it needs mapping or not and
-		 * we calculated position in bit array and size to check in bit array
-		 */
-		if (!sys_bitarray_is_region_cleared(
-				heap->physical_blocks_allocators[mem_block_iterator]->bitmap,
-				check_size, check_position))
-			continue;
-
-		/* needed in case of mid mapping failure
-		 * Rewrite of the concept is needed. Instead of failure handling we
-		 * need to check if there is enough physical memory available.
-		 */
-		if (!ptrs_to_map) {
-			ptrs_to_map = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED,
-				0, SOF_MEM_CAPS_RAM, sizeof(uintptr_t) * physical_block_count);
-			if (!ptrs_to_map)
-				goto fail;
-		}
-
-		ptrs_to_map[i] = phys_block_ptr;
-
-		if (sys_mm_drv_map_region((void *)phys_block_ptr, (uintptr_t)NULL,
-					CONFIG_MM_DRV_PAGE_SIZE, SYS_MM_MEM_PERM_RW) != 0) {
-			ptrs_to_map[i] = (uintptr_t)NULL;
-			goto fail;
-		}
+	allocation_error_code = vmh_map_region(heap->physical_blocks_allocators[mem_block_iterator],
+					       ptr, alloc_size);
+	if (allocation_error_code) {
+		sys_mem_blocks_free_contiguous(heap->physical_blocks_allocators[mem_block_iterator],
```

Comment:
isn't this already covered by line 308 above?

Reply:
Line 308 only unmaps memory, line 461 releases memory region from mem blocks.